back/engines/commercial/report_builder/app/services/report_builder/survey_response_grouper.rb

```diff
@@ -0,0 +1,180 @@
+# frozen_string_literal: true
+
+module ReportBuilder
+  class SurveyResponseGrouper
+    def initialize(phase)
+      @form = phase.custom_form || CustomForm.new(participation_context: phase)
+      @inputs = phase.ideas.published
+    end
+
+    def get_result(question_field_id)
+      question = get_question(question_field_id)
+
+      query = @inputs.select(
+        select_query(question, as: 'answer')
+      )
+
+      answers = construct_not_grouped_answers(query, question)
+
+      answers = answers.sort_by { |a| -a[:count] }
+
+      # Build response
+      build_response(answers, question, nil)
+    end
+
+    def group_by_user_field(question_field_id, user_field_id)
+      question = get_question(question_field_id)
+      user_field = CustomField.find_by(id: user_field_id)
+      throw "Unsupported user field type: #{user_field.input_type}" unless user_field.input_type == 'select'
+
+      query = @inputs.joins(:author)
+
+      query = query.select(
+        select_query(question, as: 'answer'),
+        select_query(user_field, as: 'group')
+      )
+
+      answers = construct_grouped_answers(query, question, user_field)
+
+      build_response(answers, question, user_field)
+    end
+
+    def group_by_other_question(question_field_id, other_question_field_id)
+      question = get_question(question_field_id)
+      other_question = get_question(other_question_field_id)
+      throw "Unsupported question type: #{other_question.input_type}" unless other_question.input_type == 'select'
+
+      query = @inputs.select(
+        select_query(question, as: 'answer'),
+        select_query(other_question, as: 'group')
+      )
+
+      answers = construct_grouped_answers(query, question, other_question)
+
+      build_response(answers, question, other_question)
+    end
+
+    def get_multilocs(question, group_field)
+      multilocs = { answer: get_option_titles(question) }
+      multilocs[:group] = get_option_titles(group_field) if group_field
+      multilocs
+    end
+
+    private
+
+    def get_question(question_field_id)
+      question = @form.custom_fields.find_by(id: question_field_id)
+      throw 'Question not found' unless question
+      throw "Unsupported question type: #{question.input_type}" unless %w[select multiselect].include?(question.input_type)
+
+      question
+    end
+
+    def select_query(field, as: 'answer')
+      table = field.resource_type == 'User' ? 'users' : 'ideas'
+
+      if field.input_type == 'select'
+        "#{table}.custom_field_values->'#{field.key}' as #{as}"
+      else
+        %{
+          jsonb_array_elements(
+            CASE WHEN jsonb_path_exists(#{table}.custom_field_values, '$ ? (exists (@.#{field.key}))')
```

Comment:
@jamesspeake Just to be sure we're on the same page. I believe we cannot apply your trick https://github.com/CitizenLabDotCo/citizenlab/pull/7028/files#diff-f519d3e628f000e1fe7fb51de96cb14194d9cefaa2e4bc308d781f53858c1419R137 here because we need to see **how many** responses do not include this answer.

And unfortunately, if `jsonb_array_elements` is used in `select id, jsonb_array_elements(...`  and `jsonb_array_elements` result is empty, then [the row is not returned](https://dba.stackexchange.com/a/222916) at all even if `id` is present.